vexination/walker/src/scanner.rs

```diff
@@ -0,0 +1,126 @@
+        interval.set_missed_tick_behavior(MissedTickBehavior::Skip);
+
+        loop {
+            if let Err(err) = self.run_once().await {
+                log::warn!("Failed to perform sync: {err}");
+            }
+            interval.tick().await;
+        }
+    }
+
+    #[instrument(skip(self))]
+    pub async fn run_once(&self) -> anyhow::Result<()> {
+        let since = Since::new(None::<SystemTime>, self.options.since_file.clone(), Default::default())?;
+        let source: DispatchSource = match Url::parse(&self.options.source) {
+            Ok(url) => HttpSource {
+                url,
+                fetcher: Fetcher::new(FetcherOptions::default()).await?,
+                options: csaf_walker::source::HttpOptions { since: *since },
+            }
+            .into(),
+            Err(_) => FileSource::new(&self.options.source, None)?.into(),
+        };
+
+        let sender = sender::HttpSender::new(
+            self.options.provider.clone(),
+            sender::Options {
+                additional_root_certificates: self.options.additional_root_certificates.clone(),
+                ..sender::Options::default()
+            },
+        )
+        .await?;
+
+        let storage = walker_extras::visitors::SendVisitor {
+            url: self.options.target.clone(),
+            sender,
+            retries: self.options.retries,
+            retry_delay: self.options.retry_delay,
+        };
+
+        let validation = ValidationVisitor::new(storage).with_options(ValidationOptions {
+            validation_date: self.options.validation_date,
+        });
+
+        let walker = Walker::new(source.clone());
+        walker
+            .walk(FilteringVisitor {
+                config: FilterConfig {
+                    ignored_distributions: HashSet::from_iter(self.options.ignore_distributions.clone()),
+                    ignored_prefixes: vec![],
+                    only_prefixes: self.options.required_prefixes.clone(),
+                },
+                visitor: RetrievingVisitor::new(source.clone(), validation),
+            })
+            .await?;
+
+        let sender = sender::HttpSender::new(
+            self.options.provider.clone(),
+            sender::Options {
+                additional_root_certificates: self.options.additional_root_certificates.clone(),
+                ..sender::Options::default()
+            },
+        )
+        .await?;
+
+        let storage = walker_extras::visitors::SendVisitor {
+            url: self.options.target.clone(),
+            sender,
+            retries: self.options.retries,
+            retry_delay: self.options.retry_delay,
+        };
+        let validation = ValidationVisitor::new(storage).with_options(ValidationOptions {
+            validation_date: self.options.validation_date,
+        });
+
+        let walker = Walker::new(source.clone());
+        walker.walk(RetrievingVisitor::new(source.clone(), validation)).await?;
```

Comment:
This looks like a mistake. You're creating the same variables you created above.

Reply:
I used shadowing because it seems necessary to walk the filtering and the retrieving visitor. 
Actually it might in the wrong order.
And cloning those is not possible.

Reply:
Used shadowing for what? Why call `walk` twice?